vector-im/element-android
# Replace legacy lock screen implementation

## Type of change


## Content

Our PF-LockScreen fork is replaced by our own implementation of this feature. This should be easier to customise and maintain, it allows us to use the new biometric APIs from Android, removing the deprecated ones, and adds missing features like biometric key invalidation on new method enrolled.

In the beginning this was a separate module, but due to dependencies with some classes in the vector module it had to be included as part of this module. It's in its own `im.vector.app.features.pin.lockscreen` package so it can be split in the future. Sadly, the resources needed are mostly in the `library/ui-styles` module and are not as easy to split as they could be, but most of them are prefixed using `lockscreen_` to be easily distinguished.

## Motivation and context

Fixes #6217, depends on #6270 .

## Screenshots / GIFs

No UI changes, we're using the same resources and styles.

## Tests



- Have a logged in user.
- Go to Settings -> Security & Privacy -> Protect access.
- Enable PIN, set up a PIN code.
- Enable biometrics too if possible, confirm with a fingerprint (this can also be done on emulators from the settings dialog, but emulator with API 29 had some issues with fingerprints).
- To test this feature more easily, disable "Require PIN after 2 minutes" so the PIN code screen is displayed every time you return to the app.

Which things could be useful to test:
- Enabling / disabling biometrics.
- Failing the biometric prompt too many times, it should automatically disable biometric auth.
- Failing PIN code auth too many times should force a log out.
- Changing PIN.

Note: if the user updates the app and they already had both PIN code and biometric authentication enabled, _key invalidation will only start working after the first successful biometric auth_. A system/biometric key will be created and maintained. Also, key invalidation is only available since Android 6.

## Tested devices

- OS version(s): 9, 10, 12 (both emulator and physical devices).

## Checklist
